console: Add semihosting console

LuoZhongYao opened the pull request:
```
Many chips have only one serial port. When the serial port is occupied
by other devices, there is a lack of a console to output debugging
information. Semihosting can provide a console. The disadvantage of
semihosting is that a debugger must be connected, so it can only be
used for online debugging.

Signed-off-by: ZhongYao Luo <LuoZhongYao@gmail.com>
```

LuoZhongYao committed:
console: Add semihosting console

Many chips have only one serial port. When the serial port is occupied
by other devices, there is a lack of a console to output debugging
information. Semihosting can provide a console. The disadvantage of
semihosting is that a debugger must be connected, so it can only be
used for online debugging.

Signed-off-by: ZhongYao Luo <LuoZhongYao@gmail.com>
Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

jfischer-phytec-iot commented on drivers/console/semihost_console.c:
```diff
@@ -0,0 +1,39 @@
+	__asm__ __volatile__ (
+		"MOVS	R1, %0		\n"
+		"MOVS	R0, #3		\n"
+		"BKPT	0xAB		\n"
```
`bkpt` works only for ARMv6-M and ARMv7-M

jfischer-phytec-iot commented on drivers/console/semihost_console.c:
```diff
@@ -0,0 +1,39 @@
+
+	c = ch;
+	__asm__ __volatile__ (
+		"MOVS	R1, %0		\n"
```
No reason to use capital letters. Please fix all style issues.

aescolar commented:
@LuoZhongYao The commit author is 
Author: LuoZhongYao <LuoZhongYao@gmail.com>
but the signed off line is:
Signed-off-by: ZhongYao Luo <LuoZhongYao@gmail.com>
You'll need to amend the author or signed off line to match each other

aescolar commented on drivers/console/Kconfig:
```diff
@@ -291,6 +291,12 @@ config NATIVE_POSIX_CONSOLE_INIT_PRIORITY
 
+config SEMIHOST_CONSOLE
+	bool "Use semihosting for console"
+	select CONSOLE_HAS_DRIVER
```
This kconfig option would need to `depend on` the arch it has assembler for.

stephanosio commented on drivers/console/semihost_console.c:
```diff
@@ -0,0 +1,39 @@
+	__asm__ __volatile__ (
+		"movs	r1, %0\n"
+		"movs	r0, #3\n"
+		"bkpt	0xab\n"
```
Note that this trap mechanism to specific to Cortex-M and will not work on Cortex-A and -R devices.

For Cortex-A and -R, the trap encoding is active instruction set-dependent (i.e. T32, A32 and A64 all have different encoding).

![image](https://user-images.githubusercontent.com/11345418/69901276-bcc79100-13c2-11ea-8425-d132d8ef3a59.png)
(from "Semihosting for AArch32 and AArch64 Release 2.0")

stephanosio commented on drivers/console/Kconfig:
```diff
@@ -291,6 +291,13 @@ config NATIVE_POSIX_CONSOLE_INIT_PRIORITY
+config SEMIHOST_CONSOLE
+	bool "Use semihosting for console"
+	select CONSOLE_HAS_DRIVER
+	depends on CPU_CORTEX
```
Since the semihosting trap mechanism used here is specific to Cortex-M, it should `depends on CPU_CORTEX_M`.

jfischer-no reviewed (commented):
LGTM

nashif reviewed (dismissed):
Need a test that enables this Kconfig and allows sanitycheck to build the driver, also it would be nice to have this documented somewhere to illustrate how this new feature can be used.

LuoZhongYao commented:
> Need a test that enables this Kconfig and allows sanitycheck to build the driver, also it would be nice to have this documented somewhere to illustrate how this new feature can be used.

LuoZhongYao commented:
> Need a test that enables this Kconfig and allows sanitycheck to build the driver, also it would be nice to have this documented somewhere to illustrate how this new feature can be used.

Semihosting requires a debugger to be tested, My current board needs to be tested in Keil MDK, Is there any documentation on how to write such test cases.

stephanosio commented:
@LuoZhongYao It seems qemu has `-semihosting` option, so this could potentially be used for verifying the semihosting console driver. If this is not feasible for some reason, at minimum, we could a sample that uses this driver and build it to verify that there is no build error.

stephanosio committed:
cmake: emu: Support QEMU semihosting console emulation

The QEMU '-semihosting' option enables the emulation of the semihosting
mechanism that can be used to interface the Zephyr RTOS to the host
operating system.

In order to support semihosting console output, the QEMU semihosting
feature is enabled and its console is connected to the console backend
chardev.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: drivers: Add console driver test

This commit adds the initial implementation of the console driver test.

The purpose of this test is to verify the output functionality of the
various types of console drivers (UART and semihost console types are
supported for now).

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio commented:
Added a console driver test (`tests/drivers/console`) that covers both UART and semihosting consoles.

stephanosio commented:
Rebased. Please review and merge.

LuoZhongYao commented:
> 
> 
> Rebased. Please review and merge.

me?

stephanosio commented:
@nashif Could you get back to this? Your change request is still pending.

stephanosio commented:
Ping

nashif reviewed (changes requested):
my review was not completely addressed. I am not sure exactly what this test does and what it exactly tests. Right now the first test basically prints something on the screen. How is this different from the hello world for example? The second one builds for all ARM, but are we testing that semihosting actually works? Why do we need both tests? 
Can we add some documentation on how this would work with qemu somewhere?

stephanosio commented:
> How is this different from the hello world for example?

Hello World sample simply prints out "Hello World" message, whereas this test actually checks that the printed message is properly output on different console backends.

Also, this test can be further extended to validate console inputs as well in the future.

> The second one builds for all ARM, but are we testing that semihosting actually works?

Yes, the `drivers.console.semihost` test validates that the message is properly printed through the semihosting console backend.

https://github.com/zephyrproject-rtos/zephyr/pull/21073/commits/a474b56570bbd99c0a0d0b627d4b95fe9185a82b#diff-908e1b90e51603761a8e3fa2b098edc1R56
https://github.com/zephyrproject-rtos/zephyr/pull/21073/commits/27fcc49b0e2be5e91caaffd50572687f86b61c4e#diff-591623c87f980d3ff4279dc45608bb3fR1-R7

> Why do we need both tests?

The first one tests console output through UART backend, while the second one tests console output through semihosting console backend.

stephanosio commented:
> Can we add some documentation on how this would work with qemu somewhere?

@nashif well, for QEMU, it just automatically works when `CONFIG_SEMIHOST_CONSOLE=y` (i.e. semihosting console backend is enabled).

As for real hardware, interacting with the semihosting mechanism is debugger-specific and the debugger vendor (or SoC vendor) provides documentations for it.

carlescufi commented:
@nashif, @LuoZhongYao and @stephanosio I think we should link to this page somewhere in the doc:

https://developer.arm.com/architectures/learn-the-architecture/debugger-usage/semihosting

carlescufi commented on drivers/console/Kconfig:
```diff
@@ -291,6 +291,13 @@ config NATIVE_POSIX_CONSOLE_INIT_PRIORITY
+config SEMIHOST_CONSOLE
+	bool "Use semihosting for console"
+	select CONSOLE_HAS_DRIVER
+	depends on CPU_CORTEX_M
```
Why CORTEX_M? This is supported in CORTEX_A too

stephanosio replied:
@carlescufi https://github.com/zephyrproject-rtos/zephyr/pull/21073#discussion_r352286180

The current implementation is Cortex-M-specific, and that is arguably where this feature is the most needed since low-end Cortex-M devices are often I/O constrained.

As for Cortex-A and Cortex-R, I plan to add support for them later if there is a need for it (note that the semihosting mechanism can do more than just console I/O, i.e. it can do virtual file I/O as well and this can be useful for Cortex-A and -R as well).

carlescufi replied:
Very well, thank you.

stephanosio commented:
> @nashif, @LuoZhongYao and @stephanosio I think we should link to this page somewhere in the doc:
> 
> https://developer.arm.com/architectures/learn-the-architecture/debugger-usage/semihosting

@carlescufi @nashif Where do you think would be the best place to add this?

carlescufi commented:
> > @nashif, @LuoZhongYao and @stephanosio I think we should link to this page somewhere in the doc:
> > https://developer.arm.com/architectures/learn-the-architecture/debugger-usage/semihosting
> 
> @carlescufi @nashif Where do you think would be the best place to add this?

@stephanosio 
Unfortunately there doesn't seem to be a reference page for the console. I would then suggest adding one in https://github.com/zephyrproject-rtos/zephyr/tree/master/doc/reference (perhaps console/index.rst) and then adding it there.

carlescufi commented:
@LuoZhongYao could you rebase please?

tejlmand commented on tests/drivers/console/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
```
Please switch to `find_package(Zephyr)`
```suggestion
find_package(Zephyr HINTS $ENV{ZEPHYR_BASE})
```

stephanosio replied:
Done

tejlmand reviewed (changes requested):
Please switch to `find_package(Zephyr)`

LuoZhongYao commented:
> 
> 
> @LuoZhongYao could you rebase please?

what rebase and how?

carlescufi commented:
> > @LuoZhongYao could you rebase please?
> 
> what rebase and how?

you fetch the upstream:

```
git fetch git@github.com:zephyrproject-rtos/zephyr.git
git rebase FETCH_HEAD
git push -f
```

Also can you please address the comments from @nashif and @tejlmand ?

stephanosio commented:
Rebased. Still need to address some documentation issues.

I will eventually get back to it, unless @LuoZhongYao or someone else wants to look into it.

carlescufi commented:
> Rebased. Still need to address some documentation issues.
> 
> I will eventually get back to it, unless @LuoZhongYao or someone else wants to look into it.

I'll take the rest of the changes @stephanosio, thanks.

carlescufi commented:
@nashif I have documented semihosting in the Kconfig help, since there is not a console doc page right now. See here:
https://github.com/zephyrproject-rtos/zephyr/pull/21073/files#diff-787d19abda649e2d3b5d86c7864acd14R299

@stephanosio already explained the different nature of the 2 tests, but I am unsure where the info he provided could be included in the tree.
